File: Java/lab19_Observer/task_2_1/src/WeatherReader.java
import java.io.*;
import java.util.ArrayDeque;
import java.util.Queue;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;

public class WeatherReader {

    private final Queue<WeatherEvent> eventQueue = new ArrayDeque<>();
    private final Lock lock = new ReentrantLock(true);
    private final String filePath;

    public WeatherReader(String filePath) {
        this.filePath = filePath;

        Thread thread = new Thread(this::read);
        thread.start();
    }

    public void read() {
        try (BufferedReader reader = new BufferedReader(new FileReader(filePath))) {
            reader.lines().forEach(line -> {
                        String[] s = line.split(" ");
                        if (s.length == 4) {
                            WeatherEvent event = new WeatherEvent(s[0], Float.parseFloat(s[1]), Float.parseFloat(s[2]), Float.parseFloat(s[3]));
                            lock.lock();
                            eventQueue.add(event);
                            lock.unlock();
                        }

                        try {
                            Thread.sleep(1000);
                        } catch (InterruptedException e) {
                            e.printStackTrace();
                        }
                    }
            );
        } catch (IOException e) {
            e.printStackTrace();
        } finally {
            WeatherEvent event = new WeatherEvent("end", 0, 9, 0                    );
            lock.lock();
            eventQueue.add(event);
            lock.unlock();
        }
    }

    public WeatherEvent getEvent() {
        lock.lock();
        WeatherEvent poll = eventQueue.poll();
        lock.unlock();

        return poll;
    }
}

File: Cpp/lab19_Observer/task_2_1/src/weather_reader.cpp
#include <thread>
#include <chrono>
#include <iostream>
#include <fstream>

#include "weather_reader.hpp"


void WeatherReader::read()
{
  std::string location;
  float temperature;
  float humidity;
  float pressure;

  WeatherEvent* event;

  try
  {
    std::ifstream reader(this->filePath, std::ios_base::in);
    while (reader >> location >> temperature >> humidity >> pressure)
    {
      event = new WeatherEvent(location, temperature, humidity, pressure);
      this->mtx.lock();
      this->eventQueue.push(event);
      this->mtx.unlock();

      std::this_thread::sleep_for(std::chrono::seconds(1));
    }
  }
  catch (std::string msg)
  {
    std::cerr << msg << std::endl;
  }

  event = new WeatherEvent("end", 0, 9, 0);
  this->mtx.lock();
  this->eventQueue.push(event);
  this->mtx.unlock();
}

WeatherReader::WeatherReader(std::string& filePath)
: filePath(filePath)
{
  std::thread thd(&WeatherReader::read, this);
  thd.detach();
}

WeatherReader::~WeatherReader()
{
  WeatherEvent* poll;

  while (!this->eventQueue.empty())
  {
    poll = this->eventQueue.front();
    this->eventQueue.pop();
    delete poll;
  }
}

WeatherEvent* WeatherReader::getEvent()
{
  WeatherEvent* poll(nullptr);

  this->mtx.lock();
  if (!this->eventQueue.empty())
  {
    poll = this->eventQueue.front();
    this->eventQueue.pop();
  }
  this->mtx.unlock();

  return poll;
}
